### DISCUSSION TITLE
Respect "For" Duration for NoData alerts

### DISCUSSION THREAD

**Discussion by joeblubaugh** [grafana/grafana#48849] [Alerting]
Grafana 8 alerting lets you define states for query execution errors and no data conditions. You can have them alert, have them be "OK" or have them notify you with the states "Error" and "No Data".

In the NoData condition mapping code there's a switch:
```go
func (a *State) resultNoData(alertRule *ngModels.AlertRule, result eval.Result) {
	a.Error = result.Error

	if a.StartsAt.IsZero() {
		a.StartsAt = result.EvaluatedAt
	}
	a.setEndsAt(alertRule, result)

	switch alertRule.NoDataState {
	case ngModels.Alerting:
		a.State = eval.Alerting
	case ngModels.NoData:
		a.State = eval.NoData
	case ngModels.OK:
		a.State = eval.Normal
	}
}
```
This switch will immediately set the value of an alert to Alerting on the first NoData evaluation. I guessed that it would wait for the "For" period in the same way that Error alerting does. Should it do that?

**Comment by yuri-tceretian** [comment]
The same way we [did](https://github.com/grafana/grafana/pull/47052) for option `Alerting` for Error state, this option should consider the rule's `for` parameter.

https://github.com/grafana/grafana/blob/4b417c8f3e7d77b6d790d77432bb8e60f634fdef/pkg/services/ngalert/state/state.go#L128-L138

**Reply by joeblubaugh** [reply]
I agree with you, do you know if there's any reason it should be different?

**Reply by yuri-tceretian** [reply]
No, I think we should fix that!

**Reply by grigassukys** [reply]
Is this gonna be fixed?

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
